p2p/swarm.go

```diff
@@ -1,135 +1,395 @@
+	// set in construction and immutable state
+	lNode *node.LocalNode
 
-	// Register a node with the swarm based on its id and tcp address but don't attempt to connect to it
-	RegisterNode(data node.RemoteNodeData)
+	// map between protocol names to listening protocol handlers
+	// NOTE: maybe let more than one handler register on a protocol ?
+	protocolHandlers     map[string]chan service.Message
+	protocolHandlerMutex sync.RWMutex
 
-	// Attempt to establish a session with a remote node with a known id and tcp address
-	// Used for bootstrapping known bootstrap nodes
-	ConnectTo(req node.RemoteNodeData, done chan error)
+	network net.Net
 
-	// Connect to count random nodes - used for bootstrapping the swarm
-	ConnectToRandomNodes(count int)
+	cPool *connectionpool.ConnectionPool
 
-	// Forcefully disconnect form a node - close any connections and sessions with it
-	DisconnectFrom(req node.RemoteNodeData)
+	dht *dht.DHT
 
-	// Send a handshake protocol message that is used to establish a session
-	sendHandshakeMessage(req SendMessageReq)
+	// Shutdown the loop
+	shutdown chan struct{} // local request to kill the swarm from outside. e.g when local node is shutting down
+}
 
-	// Register a callback channel for state changes related to remote nodes
-	// Currently used for testing network bootstrapping
-	RegisterNodeEventsCallback(callback NodeEventCallback)
+// newSwarm creates a new P2P instance
+func newSwarm(config config.Config, loadIdentity bool) (*swarm, error) {
 
-	// todo: allow client to register callback to get notified when swarm has shut down
-	Shutdown()
+	port := config.TCPPort
+	address := fmt.Sprintf("0.0.0.0:%d", port)
 
-	// services getters
-	GetDemuxer() Demuxer
-	GetLocalNode() LocalNode
+	var l *node.LocalNode
+	var err error
+	// Load an existing identity from file if exists.
 
-	getRoutingTable() table.RoutingTable
+	if loadIdentity {
+		l, err = node.NewLocalNode(config, address, true)
+	} else {
+		l, err = node.NewNodeIdentity(config, address, true)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a node, err: %v", err)
+	}
 
-	// used by protocols and for testing
-	getHandshakeProtocol() HandshakeProtocol
-	getFindNodeProtocol() FindNodeProtocol
+	n, err := net.NewNet(config, l)
+	if err != nil {
+		return nil, fmt.Errorf("can't create swarm without a network, err: %v", err)
+	}
 
-	getPeerByConnection(connID string) Peer
+	s := &swarm{
+		config:           config,
+		lNode:            l,
+		protocolHandlers: make(map[string]chan service.Message),
+		network:          n,
+		cPool:            connectionpool.NewConnectionPool(n, l.PublicKey()),
+		shutdown:         make(chan struct{}), // non-buffered so requests to shutdown block until swarm is shut down
 
-	WaitForBootstrap() error
+	}
+
+	s.dht = dht.New(l, config.SwarmConfig, s)
+
+	s.lNode.Debug("Created swarm for local node %s, %s", l.Address(), l.Pretty())
+
+	go s.listenToNetworkMessages()
+
+	if config.SwarmConfig.Bootstrap {
```

Comment:
consider refactoring this to not using config for UT

Reply:
I added a DHTMock that can be used instead and then you can use both config and the mock. though bootstrap needs to create a DHT for itself so the only option I see here is to pass factory which we don't really want. any advice ?